### ISSUE TITLE
[BUG] Types are not propagated correctly when instantiating polymorphic operators that use folds 

### ISSUE THREAD

**Issue by Kukovec** [informalsystems/apalache#1085]
## Description

After instantiating a polymorphic operator, which uses a fold, the annotations that are propagated for the fold operator body are not correctly unified with the types derived from the operator arguments at the call site, resulting in the following error:
```
[test.tla:9:32-9:32]: Found a polymorphic type: (a79 -> Int)      E@18:24:28.109
Probable causes: an empty set { } needs a type annotation or an incorrect record field is used E@18:24:28.109
Typing input error: [test.tla:9:32-9:32]: Found a polymorphic type: (a79 -> Int) E@18:24:28.111
```

## Input specification

```
---------- MODULE test ----------

EXTENDS Integers, Apalache

\* @type: (Set(a)) => a -> Int;
PolyOper(S) == 
  LET
    \* @type: (a -> Int, a) => a -> Int;
    Add(f,e) == [ x \in DOMAIN f \union {e} |-> 0 ]
  IN FoldSet( Add, [x \in {} |-> 0], S )

Init == TRUE

Next == TRUE

Inv == PolyOper({1}) = [x \in {1} |-> 0]

====================
```

## The command line parameters used to run the tool

`--inv=Inv --length=1`

## Expected behavior

No type error.

## Additional context

This is what the last pass before the 2nd snowcat pass outputs:
```
---------------------------- MODULE 11_OutAnalysis ----------------------------

EXTENDS Integers, Sequences, FiniteSets, TLC, Apalache

(*
  @type: (() => Bool);
*)
Init_si_0000 == TRUE

(*
  @type: (() => Bool);
*)
Next_si_0000 == TRUE

(*
  @type: Bool;
*)
VCInv_si_0 ==
  FoldSet(LET (*
    @type: (((a45 -> Int), a45) => (a45 -> Int));
  *)
  t_2_si_1(f_si_1, e_si_1) == [ x$1 \in DOMAIN f$1 \union {e$1} |-> 0 ]
  IN
  t_2$1, [ x$2 \in {} |-> 0 ], {1})
    = [ x$3 \in {1} |-> 0 ]

(*
  @type: Bool;
*)
VCNotInv_si_0 ==
  ~(FoldSet(LET (*
    @type: (((a45 -> Int), a45) => (a45 -> Int));
  *)
  t_2_si_2(f_si_2, e_si_2) == [ x$4 \in DOMAIN f$2 \union {e$2} |-> 0 ]
  IN
  t_2$2, [ x$5 \in {} |-> 0 ], {1})
    = [ x$6 \in {1} |-> 0 ])

================================================================================

```
Inside `FoldSet`, the LET-defined operator is still marked as having a polymorphic type (`    @type: (((a45 -> Int), a45) => (a45 -> Int));`, however, the unification at the call site should have determined `a45 = Int`.

**Comment by shonfeder** [comment]
Is this closed by #359?

**Comment by konnov** [comment]
Have a look at #1104. Your test goes through there.
